### PR TITLE
Refactor remove token

### PR DESCRIPTION
**Pull request by nventuro**
The refactor is relatively simply, since we're just removing stuff. I also adjusted the tests (best reviewed ignoring whitespace changes).

I ended up disallowing add/remove while uninitialized for simplicity. Am considering also disallowing burning from the zero address, so that the no re-init invariant is kept.

I had to remove a test about remove token collecting fees (which I imagine it did because it used to be an exit?). At any rate, we can review add/remove and fee interations later.

### CODE REVIEW

File: pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol [hunk 1]
```diff
@@ -853,75 +855,58 @@ abstract contract ManagedPoolSettings is BasePool, ProtocolFeeCache, ReentrancyG
...
-        // Exit the pool, returning the full balance of the token to the recipient
-        (IERC20[] memory tokens, uint256[] memory unscaledBalances, ) = getVault().getPoolTokens(getPoolId());
+        // Before this function is called, the caller must have withdrawn all balance for `token` from the Pool. This
+        // means that the Pool is in an invalid state, since among other things the invariant is zero. Because we're not
+        // in a valid state and all value-changing operations will revert, we are free to modify the Pool state (e.g.
+        // alter weights).
+        // We don't need to test the zero balance since the Vault will simply revert on deregistration if this is not
+        // the case.
+
+        // Removing a token will cause for the weights of all other tokens to increase. This is fine, as there is no
```
Comment (EndymionJkb) [inline pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol:878]:
```suggestion
        // Removing a token will cause the weights of all other tokens to increase. This is fine, as there is no
```

Comment (EndymionJkb) [reply pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol:878]: We could mention there is a maximum *total* weight, even if there isn't a maximum per-token weight.

File: pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol [hunk 2]
```diff
@@ -853,75 +855,58 @@ abstract contract ManagedPoolSettings is BasePool, ProtocolFeeCache, ReentrancyG
...
     function removeToken(
         IERC20 token,
-        address recipient,
-        uint256 burnAmount,
-        uint256 minAmountOut
-    ) external authenticate nonReentrant whenNotPaused returns (uint256) {
-        // We require the pool to be initialized (shown by the total supply being nonzero) in order to remove a token,
-        // maintaining the behaviour that no exits can occur before the pool has been initialized.
-        // This prevents the AUM fee calculation being triggered before the pool contains any assets.
+        address sender,
```
Comment (TomAFrench) [inline pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol:862]: This allows the manager to burn BPT from unrelated addresses which is why we hardcoded this to `msg.sender`.

File: pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol [hunk 3]
```diff
@@ -853,75 +855,58 @@ abstract contract ManagedPoolSettings is BasePool, ProtocolFeeCache, ReentrancyG
...
+        // the case.
+
+        // Removing a token will cause for the weights of all other tokens to increase. This is fine, as there is no
+        // maximum weight. We also don't need to check that the new token exists in the Pool, as the Vault will simply
+        // revert if we try to deregister a token that is not registered.
+        // We do, however, want to check that the Pool will end up with at least two tokens. This simplifies some
+        // assumptions made elsewhere (e.g. the denormalized weight sum will always be non-zero), and doesn't greatly
+        // restrict the controller.
+
+        (IERC20[] memory tokens, , ) = getVault().getPoolTokens(getPoolId());
```
Comment (TomAFrench) [inline pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol:885]: There may be a cheaper way to check the length of this array. scaling factors perhaps?

File: pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol (hunk 2 above)
Comment (nventuro) [reply pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol:862]: They can also mint to any address via `addToken`, which is the same as they can dillute everyone. They can also add bad tokens and drain the pool. I don't think this changes the trust model in any way.

File: pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol (hunk 3 above)
Comment (nventuro) [reply pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol:885]: We don't track any length ourselves - the scaling factors are stored in mapping entries. This is because we favor 1 SLOAD access in swaps (where we know the addresses of the tokens involved), at the cost of having to fetch the tokens array in other operations such as joins.

File: pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol [hunk 4]
```diff
@@ -853,75 +855,58 @@ abstract contract ManagedPoolSettings is BasePool, ProtocolFeeCache, ReentrancyG
...
-        uint256 burnAmount,
-        uint256 minAmountOut
-    ) external authenticate nonReentrant whenNotPaused returns (uint256) {
-        // We require the pool to be initialized (shown by the total supply being nonzero) in order to remove a token,
-        // maintaining the behaviour that no exits can occur before the pool has been initialized.
-        // This prevents the AUM fee calculation being triggered before the pool contains any assets.
+        address sender,
+        uint256 burnAmount
+    ) external authenticate nonReentrant whenNotPaused {
         _require(totalSupply() > 0, Errors.UNINITIALIZED);
```
Comment (TomAFrench) [inline pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol:865]: We can remove this condition as removing a token doesn't result in an exit anymore.

File: pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol [hunk 5]
```diff
@@ -775,6 +775,8 @@ abstract contract ManagedPoolSettings is BasePool, ProtocolFeeCache, ReentrancyG
         uint256 mintAmount,
         address recipient
     ) external authenticate whenNotPaused {
+        _require(totalSupply() > 0, Errors.UNINITIALIZED);
```
Comment (TomAFrench) [inline pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol:778]: There's no need for this imo.

File: pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol (hunk 3 above)
Comment (TomAFrench) [reply pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol:885]: Yeah, forgot that _scalingFactors begins with this exact query.

File: pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol (hunk 2 above)
Comment (TomAFrench) [reply pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol:862]: It's a question about blanket vs targeted actions. Similar to DAI vs USDC, if you hold DAI then Circle could blacklist Maker and make any USDC-backed DAI worthless but they can't blacklist your wallet specifically.

Comment (TomAFrench) [reply pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol:862]: We agreed on this being an issue for the controller rather than the pool.

Comment (nventuro) [reply pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol:862]: We decided that this is ultimately a controller concern - how they use burn amount and sender is very much up to them, and part of their trust model.

**Review by TomAFrench — APPROVED** [review]
Happy to merge this once the check preventing burning tokens from the zero address is in.

**Comment by TomAFrench** [comment]
There's also some inconsistency in argument ordering between adding/removing tokens but that's not a blocker.
